File: go/generic/genMonteCarlo.go
package generic

import (
	"algos/helpers"
	"fmt"
	"math"
)

// integrate performs the Monte Carlo integration
func integrate[T interface {
	IField[T]
	IOrdered[T]
}](t T, numSamples int) float64 {
	rand := helpers.NewLCG(12345, 1345, 16645, 1013904)
	underCurve := 0
	for count := 0; count < numSamples; count++ {
		x := t.coerceFromFloat(rand.NextDouble())
		y := t.coerceFromFloat(rand.NextDouble())

		// Check if x^2 + y^2 <= 1
		if x.m(x).a(y.m(y)).le(t.one()) {
			underCurve++
		}
	}

	// Return the result: (underCurve / numSamples) * 4
	return float64(underCurve) / float64(numSamples) * 4.0
}

func TestGenMonteCarlo(mode int, numSamples int) {

	if mode == 0 {
		// Use DoubleField
		t := DoubleField{Value: 0.0}
		pi := integrate(t, numSamples)
		fmt.Printf("Go generic doublefield monte carlo\n")
		fmt.Printf("Pi is approximately: %.6f\n", pi)
		fmt.Printf("Num samples: %d\n", numSamples)
		fmt.Printf("RMS Error: %.6f\n", math.Abs(math.Pi-pi))
	} else if mode == 1 {
		// Use SingleField
		t := SingleField{Value: 0}
		pi := integrate(t, numSamples)
		fmt.Printf("Go generic singlefield monte carlo\n")
		fmt.Printf("Pi is approximately: %.6f\n", pi)
		fmt.Printf("Num samples: %d\n", numSamples)
		fmt.Printf("RMS Error: %.6f\n", math.Abs(math.Pi-pi))
	} else {
		t := IntModP{Value: 0}
		SetModulus(1000003) // Set a large prime modulus
		pi := integrate(t, numSamples)
		fmt.Printf("Go generic intmodp monte carlo\n")
		fmt.Printf("Pi is approximately: %.6f\n", pi)
		fmt.Printf("Num samples: %d\n", numSamples)
		fmt.Printf("RMS Error: %.6f\n", math.Abs(math.Pi-pi))
	}
}

File: rust/src/GenMonteCarlo.rs
use rust::helpers::lcg::Lcg;
use crate::generic::double_field::DoubleField;
use crate::generic::single_field::SingleField;
use crate::generic::int_mod_p::IntModP;
use crate::generic::int_mod_p::set_modulus;
use crate::generic::i_field::IField;
use crate::generic::i_ordered::IOrdered;
pub mod generic;

fn integrate<T: IField + IOrdered>(t: &T, num_samples: usize) -> f64 {
    let mut rng = Lcg::new(12345, 1345, 16645, 1013904);
    let mut under_curve = 0;
    for _ in 0..num_samples {
        let x = t.coerce(rng.next_double());
        let y = t.coerce(rng.next_double());
        if x.m(&x).a(&y.m(&y)).le(&t.one()) {
            under_curve += 1;
        }
    }
    (under_curve as f64 / num_samples as f64) * 4.0
}

fn main() {
    // arg1 = num_samples
    // arg2 = mode (1=SingleField, 2=DoubleField, else IntModP)
    let args: Vec<String> = std::env::args().collect();
    let mut num_samples = 1_000_000;
    let mut mode = 0; // 1 for SingleField, else for DoubleField. Don't think IntModP makes much sense here
    let prime = 1_000_000_007;
    set_modulus(prime);
    if args.len() > 1 {
        num_samples = args[1].parse().unwrap_or(1000000);
    }
    if args.len() > 2 {
        mode = args[2].parse().unwrap_or(2);
    }
    let pi;
    if(mode == 1) {
        let temp = SingleField::new(0.0);
        pi = integrate(&temp, num_samples);
    }
    else if (mode == 2) {
        let temp = DoubleField::new(0.0);
        pi = integrate(&temp, num_samples);
    }
    else {
        let temp = IntModP::new(0);
        pi = integrate(&temp, num_samples);
    }
    println!("Pi is approximately: {}", pi);
    println!("Num samples: {}", num_samples);
    println!("RMS Error: {}", (std::f64::consts::PI - pi).abs());
}